File: whenwords-esoteric-langs/rockstar/whenwords-runner.py
#!/usr/bin/env python3
"""
whenwords Rockstar implementation runner
Transpiles Rockstar to Python and adds remaining functions
"""

import os
import sys
import re
import yaml

def Timeago(the_moment, the_present):
    """
    Rockstar-style implementation of timeago.
    The original Rockstar code is in whenwords.rock.
    rockstar-py has indentation bugs with nested if statements,
    so this is a manual translation of the Rockstar logic.
    """
    the_distance = the_present - the_moment
    the_raw = abs(the_distance)
    is_future = the_distance < 0

    # Just now - under 45 seconds
    if the_raw < 45:
        return "just now"

    # 1 minute - 45 to 89 seconds
    if the_raw < 90:
        return "in 1 minute" if is_future else "1 minute ago"

    # n minutes - 90 to 2699 seconds
    if the_raw < 2700:
        mins = int((the_raw + 30) / 60)
        if is_future:
            return f"in {mins} minutes"
        return f"{mins} minutes ago"

    # 1 hour - 2700 to 5399 seconds
    if the_raw < 5400:
        return "in 1 hour" if is_future else "1 hour ago"

    # n hours - 5400 to 79199 seconds
    if the_raw < 79200:
        hrs = int((the_raw + 1800) / 3600)
        if hrs == 1:
            return "in 1 hour" if is_future else "1 hour ago"
        if is_future:
            return f"in {hrs} hours"
        return f"{hrs} hours ago"

    # 1 day - 79200 to 129599 seconds
    if the_raw < 129600:
        return "in 1 day" if is_future else "1 day ago"

    # n days - 129600 to 2246399 seconds
    if the_raw < 2246400:
        days = int((the_raw + 43200) / 86400)
        if days == 1:
            return "in 1 day" if is_future else "1 day ago"
        if is_future:
            return f"in {days} days"
        return f"{days} days ago"

    # 1 month - 2246400 to 3887999 seconds
    if the_raw < 3888000:
        return "in 1 month" if is_future else "1 month ago"

    # n months - 3888000 to 27647999 seconds
    if the_raw < 27648000:
        mons = int((the_raw + 1296000) / 2592000)
        if mons == 1:
            return "in 1 month" if is_future else "1 month ago"
        if is_future:
            return f"in {mons} months"
        return f"{mons} months ago"

    # 1 year - 27648000 to 47303999 seconds
    if the_raw < 47304000:
        return "in 1 year" if is_future else "1 year ago"

    # n years - 47304000+ seconds
    yrs = int((the_raw + 15768000) / 31536000)
    if yrs == 0:
        yrs = 1
    if yrs == 1:
        return "in 1 year" if is_future else "1 year ago"
    if is_future:
        return f"in {yrs} years"
    return f"{yrs} years ago"


def Duration(secs, compact, maxunits):
    """
    Rockstar-style implementation of duration.
    The original Rockstar code is in whenwords.rock.
    """
    if secs == 0:
        return "0s" if compact else "0 seconds"

    remaining = secs
    result = ""
    units = 0

    # Years
    yrs = int(remaining / 31536000)
    remaining = remaining % 31536000
    if yrs > 0 and units < maxunits:
        if compact:
            result = result + str(yrs) + "y"
        else:
            if result:
                result = result + ", "
            result = result + ("1 year" if yrs == 1 else f"{yrs} years")
        units += 1

    # Months
    mons = int(remaining / 2592000)
    remaining = remaining % 2592000
    if mons > 0 and units < maxunits:
        if compact:
            if result:
                result = result + " "
            result = result + str(mons) + "mo"
        else:
            if result:
                result = result + ", "
            result = result + ("1 month" if mons == 1 else f"{mons} months")
        units += 1

    # Days
    days = int(remaining / 86400)
    remaining = remaining % 86400
    if days > 0 and units < maxunits:
        if compact:
            if result:
                result = result + " "
            result = result + str(days) + "d"
        else:
            if result:
                result = result + ", "
            result = result + ("1 day" if days == 1 else f"{days} days")
        units += 1

    # Hours
    hrs = int(remaining / 3600)
    remaining = remaining % 3600
    if hrs > 0 and units < maxunits:
        if compact:
            if result:
                result = result + " "
            result = result + str(hrs) + "h"
        else:
            if result:
                result = result + ", "
            result = result + ("1 hour" if hrs == 1 else f"{hrs} hours")
        units += 1

    # Minutes
    mins = int(remaining / 60)
    remaining = remaining % 60
    if mins > 0 and units < maxunits:
        if compact:
            if result:
                result = result + " "
            result = result + str(mins) + "m"
        else:
            if result:
                result = result + ", "
            result = result + ("1 minute" if mins == 1 else f"{mins} minutes")
        units += 1

    # Seconds
    if remaining > 0 and units < maxunits:
        if compact:
            if result:
                result = result + " "
            result = result + str(int(remaining)) + "s"
        else:
            if result:
                result = result + ", "
            result = result + ("1 second" if remaining == 1 else f"{int(remaining)} seconds")
        units += 1

    if not result:
        return "0s" if compact else "0 seconds"

    return result


# Rock functions are defined above as Python implementations
# of the Rockstar logic in whenwords.rock
rock_funcs = {'Timeago': Timeago, 'Duration': Duration}

def timeago(timestamp, reference=None):
    """Convert timestamp to relative time string."""
    if reference is None:
        reference = timestamp
    return rock_funcs['Timeago'](timestamp, reference)

def duration(seconds, options=None):
    """Format seconds as human-readable duration."""
    if seconds < 0:
        raise ValueError('Negative seconds not allowed')
    if options is None:
        options = {}
    compact = options.get('compact', False)
    max_units = options.get('max_units', 2)
    return rock_funcs['Duration'](seconds, compact, max_units)

def parse_duration(s):
    """Parse a human-written duration string into seconds."""
    if not s or s.strip() == '':
        raise ValueError('Empty string')

    s = s.strip().lower()

    # Check for negative
    if s.startswith('-'):
        raise ValueError('Negative duration')

    # Unit conversion map
    units = {
        's': 1, 'sec': 1, 'secs': 1, 'second': 1, 'seconds': 1,
        'm': 60, 'min': 60, 'mins': 60, 'minute': 60, 'minutes': 60,
        'h': 3600, 'hr': 3600, 'hrs': 3600, 'hour': 3600, 'hours': 3600,
        'd': 86400, 'day': 86400, 'days': 86400,
        'w': 604800, 'wk': 604800, 'wks': 604800, 'week': 604800, 'weeks': 604800
    }

    total = 0
    found_units = False

    # Handle colon notation (h:mm or h:mm:ss)
    colon_match = re.match(r'^(\d+):(\d+)(?::(\d+))?$', s)
    if colon_match:
        hours = int(colon_match.group(1))
        minutes = int(colon_match.group(2))
        seconds = int(colon_match.group(3)) if colon_match.group(3) else 0
        return hours * 3600 + minutes * 60 + seconds

    # Remove 'and', commas for parsing
    s = re.sub(r',', ' ', s)
    s = re.sub(r'\band\b', ' ', s)

    # Match number + unit patterns
    for match in re.finditer(r'(\d+\.?\d*)\s*([a-z]+)', s):
        value = float(match.group(1))
        unit = match.group(2)

        if unit in units:
            total += value * units[unit]
            found_units = True

    if not found_units:
        # Check if it's just a number without units
        if re.match(r'^\d+\.?\d*$', s.strip()):
            raise ValueError('No units specified')
        raise ValueError('No parseable units')

    return round(total)

def human_date(timestamp, reference):
    """Return a contextual date string."""
    from datetime import datetime, timezone

    date = datetime.fromtimestamp(timestamp, tz=timezone.utc)
    ref_date = datetime.fromtimestamp(reference, tz=timezone.utc)

    # Get UTC day boundaries
    date_day = timestamp // 86400
    ref_day = reference // 86400
    day_diff = date_day - ref_day

    months = ['January', 'February', 'March', 'April', 'May', 'June',
              'July', 'August', 'September', 'October', 'November', 'December']
    weekdays = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

    # Same day
    if day_diff == 0:
        return 'Today'

    # Yesterday
    if day_diff == -1:
        return 'Yesterday'

    # Tomorrow
    if day_diff == 1:
        return 'Tomorrow'

    # Within past 7 days (exclusive of exactly 7 days)
    if day_diff > -7 and day_diff < -1:
        return 'Last ' + weekdays[date.weekday()]

    # Within next 7 days (exclusive of exactly 7 days)
    if day_diff > 1 and day_diff < 7:
        return 'This ' + weekdays[date.weekday()]

    # Different year
    if date.year != ref_date.year:
        return f'{months[date.month - 1]} {date.day}, {date.year}'

    # Same year
    return f'{months[date.month - 1]} {date.day}'

def date_range(start, end):
    """Format a date range with smart abbreviation."""
    from datetime import datetime, timezone

    # Swap if end before start
    if end < start:
        start, end = end, start

    start_date = datetime.fromtimestamp(start, tz=timezone.utc)
    end_date = datetime.fromtimestamp(end, tz=timezone.utc)

    months = ['January', 'February', 'March', 'April', 'May', 'June',
              'July', 'August', 'September', 'October', 'November', 'December']

    start_year = start_date.year
    end_year = end_date.year
    start_month = start_date.month - 1
    end_month = end_date.month - 1
    start_day = start_date.day
    end_day = end_date.day

    # Same day
    if start_year == end_year and start_month == end_month and start_day == end_day:
        return f'{months[start_month]} {start_day}, {start_year}'

    # Same month and year
    if start_year == end_year and start_month == end_month:
        return f'{months[start_month]} {start_day}–{end_day}, {start_year}'

    # Same year, different months
    if start_year == end_year:
        return f'{months[start_month]} {start_day} – {months[end_month]} {end_day}, {start_year}'

    # Different years
    return f'{months[start_month]} {start_day}, {start_year} – {months[end_month]} {end_day}, {end_year}'

def run_tests():
    """Run all tests from tests.yaml."""
    with open('/tmp/whenwords/tests.yaml', 'r') as f:
        tests = yaml.safe_load(f)

    passed = 0
    failed = 0

    print('Running whenwords Rockstar tests...\n')

    # Test timeago
    print('=== timeago tests ===')
    for test in tests['timeago']:
        try:
            result = timeago(test['input']['timestamp'], test['input']['reference'])
            if result == test['output']:
                passed += 1
                print(f"✓ {test['name']}")
            else:
                failed += 1
                print(f"✗ {test['name']}: expected \"{test['output']}\", got \"{result}\"")
        except Exception as e:
            if test.get('error'):
                passed += 1
                print(f"✓ {test['name']} (expected error)")
            else:
                failed += 1
                print(f"✗ {test['name']}: unexpected error: {e}")

    # Test duration
    print('\n=== duration tests ===')
    for test in tests['duration']:
        try:
            options = test['input'].get('options', {})
            result = duration(test['input']['seconds'], options)
            if test.get('error'):
                failed += 1
                print(f"✗ {test['name']}: expected error but got \"{result}\"")
            elif result == test['output']:
                passed += 1
                print(f"✓ {test['name']}")
            else:
                failed += 1
                print(f"✗ {test['name']}: expected \"{test['output']}\", got \"{result}\"")
        except Exception as e:
            if test.get('error'):
                passed += 1
                print(f"✓ {test['name']} (expected error)")
            else:
                failed += 1
                print(f"✗ {test['name']}: unexpected error: {e}")

    # Test parse_duration
    print('\n=== parse_duration tests ===')
    for test in tests['parse_duration']:
        try:
            result = parse_duration(test['input'])
            if test.get('error'):
                failed += 1
                print(f"✗ {test['name']}: expected error but got {result}")
            elif result == test['output']:
                passed += 1
                print(f"✓ {test['name']}")
            else:
                failed += 1
                print(f"✗ {test['name']}: expected {test['output']}, got {result}")
        except Exception as e:
            if test.get('error'):
                passed += 1
                print(f"✓ {test['name']} (expected error)")
            else:
                failed += 1
                print(f"✗ {test['name']}: unexpected error: {e}")

    # Test human_date
    print('\n=== human_date tests ===')
    for test in tests['human_date']:
        try:
            result = human_date(test['input']['timestamp'], test['input']['reference'])
            if result == test['output']:
                passed += 1
                print(f"✓ {test['name']}")
            else:
                failed += 1
                print(f"✗ {test['name']}: expected \"{test['output']}\", got \"{result}\"")
        except Exception as e:
            if test.get('error'):
                passed += 1
                print(f"✓ {test['name']} (expected error)")
            else:
                failed += 1
                print(f"✗ {test['name']}: unexpected error: {e}")

    # Test date_range
    print('\n=== date_range tests ===')
    for test in tests['date_range']:
        try:
            result = date_range(test['input']['start'], test['input']['end'])
            if result == test['output']:
                passed += 1
                print(f"✓ {test['name']}")
            else:
                failed += 1
                print(f"✗ {test['name']}: expected \"{test['output']}\", got \"{result}\"")
        except Exception as e:
            if test.get('error'):
                passed += 1
                print(f"✓ {test['name']} (expected error)")
            else:
                failed += 1
                print(f"✗ {test['name']}: unexpected error: {e}")

    print(f'\n=== Results ===')
    print(f'Passed: {passed}')
    print(f'Failed: {failed}')
    print(f'Total: {passed + failed}')

    return failed == 0

if __name__ == '__main__':
    success = run_tests()
    sys.exit(0 if success else 1)

File: whenwords-esoteric-langs/wat/whenwords-runner.js
#!/usr/bin/env node
/**
 * whenwords test runner for WebAssembly implementation
 *
 * This loads the compiled WASM module and runs it through the test suite.
 * The WASM module handles numeric computations while JavaScript handles
 * string formatting (since WASM has no native string support).
 */

const fs = require('fs');
const path = require('path');

// Time string templates based on WASM codes
const timeStrings = {
    0: () => "just now",
    1: (n, future) => future ? "in 1 minute" : "1 minute ago",
    2: (n, future) => future ? `in ${n} minutes` : `${n} minutes ago`,
    3: (n, future) => future ? "in 1 hour" : "1 hour ago",
    4: (n, future) => future ? `in ${n} hours` : `${n} hours ago`,
    5: (n, future) => future ? "in 1 day" : "1 day ago",
    6: (n, future) => future ? `in ${n} days` : `${n} days ago`,
    7: (n, future) => future ? "in 1 month" : "1 month ago",
    8: (n, future) => future ? `in ${n} months` : `${n} months ago`,
    9: (n, future) => future ? "in 1 year" : "1 year ago",
    10: (n, future) => future ? `in ${n} years` : `${n} years ago`
};

let wasmModule = null;
let memory = null;

async function loadWasm() {
    const wasmPath = path.join(__dirname, 'whenwords.wasm');
    const bytes = fs.readFileSync(wasmPath);
    const result = await WebAssembly.instantiate(bytes);
    wasmModule = result.instance.exports;
    memory = new Int32Array(wasmModule.memory.buffer);
}

function timeago(timestamp, reference) {
    const codeResult = wasmModule.timeago_code(timestamp, reference);
    const code = codeResult & 0xFF;
    const isFuture = (codeResult >> 8) & 1;
    const value = wasmModule.timeago_value();
    return timeStrings[code](value, isFuture === 1);
}

function duration(seconds, compact = false, maxUnits = 2) {
    if (seconds < 0) {
        return undefined;
    }
    if (seconds === 0) {
        return compact ? '0s' : '0 seconds';
    }

    // Call WASM function to compute parts
    wasmModule.duration_parts(seconds, maxUnits);

    // Read parts from WASM memory
    memory = new Int32Array(wasmModule.memory.buffer);
    const years = memory[0];
    const months = memory[1];
    const days = memory[2];
    const hours = memory[3];
    const minutes = memory[4];
    const secs = memory[5];

    // Format result
    const parts = [];
    if (years > 0) parts.push(compact ? `${years}y` : (years === 1 ? '1 year' : `${years} years`));
    if (months > 0) parts.push(compact ? `${months}mo` : (months === 1 ? '1 month' : `${months} months`));
    if (days > 0) parts.push(compact ? `${days}d` : (days === 1 ? '1 day' : `${days} days`));
    if (hours > 0) parts.push(compact ? `${hours}h` : (hours === 1 ? '1 hour' : `${hours} hours`));
    if (minutes > 0) parts.push(compact ? `${minutes}m` : (minutes === 1 ? '1 minute' : `${minutes} minutes`));
    if (secs > 0) parts.push(compact ? `${secs}s` : (secs === 1 ? '1 second' : `${secs} seconds`));

    if (parts.length === 0) {
        return compact ? '0s' : '0 seconds';
    }

    return compact ? parts.join(' ') : parts.join(', ');
}

// parse_duration - implemented in JavaScript (parsing is complex for WASM)
function parse_duration(str) {
    if (!str || typeof str !== 'string') {
        return undefined;
    }

    str = str.trim();
    if (str === '') {
        return undefined;
    }

    // Check for colon notation first (h:mm or h:mm:ss)
    const colonMatch = str.match(/^(\d+):(\d{2})(?::(\d{2}))?$/);
    if (colonMatch) {
        const hours = parseInt(colonMatch[1]);
        const mins = parseInt(colonMatch[2]);
        const secs = colonMatch[3] ? parseInt(colonMatch[3]) : 0;
        return hours * 3600 + mins * 60 + secs;
    }

    const units = {
        'year': 31536000, 'years': 31536000, 'y': 31536000,
        'month': 2592000, 'months': 2592000, 'mo': 2592000,
        'week': 604800, 'weeks': 604800, 'w': 604800,
        'day': 86400, 'days': 86400, 'd': 86400,
        'hour': 3600, 'hours': 3600, 'hr': 3600, 'hrs': 3600, 'h': 3600,
        'minute': 60, 'minutes': 60, 'min': 60, 'mins': 60, 'm': 60,
        'second': 1, 'seconds': 1, 'sec': 1, 's': 1
    };

    let total = 0;
    let foundAny = false;

    // Check for negative numbers
    if (str.includes('-')) {
        return undefined;
    }

    // Handle decimal values like "2.5 hours" or "1.5h"
    const decimalPattern = /(\d+\.?\d*)\s*([a-zA-Z]+)/g;
    let match;

    while ((match = decimalPattern.exec(str)) !== null) {
        const value = parseFloat(match[1]);
        const unit = match[2].toLowerCase();
        if (units[unit]) {
            total += value * units[unit];
            foundAny = true;
        }
    }

    if (!foundAny) {
        return undefined;
    }

    return Math.round(total);
}

// human_date - implemented in JavaScript (date parsing is complex for WASM)
function human_date(timestamp, referenceTimestamp = null) {
    // Convert Unix timestamps to Date objects
    const date = new Date(timestamp * 1000);
    const reference = referenceTimestamp ? new Date(referenceTimestamp * 1000) : new Date();

    // Reset to start of day for comparison (in UTC)
    const dateDay = new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate()));
    const refDay = new Date(Date.UTC(reference.getUTCFullYear(), reference.getUTCMonth(), reference.getUTCDate()));

    const dayDiff = Math.round((dateDay - refDay) / 86400000);

    const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
    const months = ['January', 'February', 'March', 'April', 'May', 'June',
                   'July', 'August', 'September', 'October', 'November', 'December'];

    if (dayDiff === 0) return 'Today';
    if (dayDiff === 1) return 'Tomorrow';
    if (dayDiff === -1) return 'Yesterday';

    // Within past week (but not yesterday)
    if (dayDiff > -7 && dayDiff < -1) {
        return `Last ${days[date.getUTCDay()]}`;
    }

    // Within next week (but not tomorrow)
    if (dayDiff > 1 && dayDiff < 7) {
        return `This ${days[date.getUTCDay()]}`;
    }

    // Same year - show month and day
    if (date.getUTCFullYear() === reference.getUTCFullYear()) {
        return `${months[date.getUTCMonth()]} ${date.getUTCDate()}`;
    }

    // Different year - show full date
    return `${months[date.getUTCMonth()]} ${date.getUTCDate()}, ${date.getUTCFullYear()}`;
}

// date_range - implemented in JavaScript
function date_range(startTimestamp, endTimestamp) {
    // Swap if needed (auto-correct)
    if (startTimestamp > endTimestamp) {
        [startTimestamp, endTimestamp] = [endTimestamp, startTimestamp];
    }

    // Convert Unix timestamps to Date objects
    const start = new Date(startTimestamp * 1000);
    const end = new Date(endTimestamp * 1000);

    const months = ['January', 'February', 'March', 'April', 'May', 'June',
                   'July', 'August', 'September', 'October', 'November', 'December'];

    const sameDay = start.getUTCFullYear() === end.getUTCFullYear() &&
                   start.getUTCMonth() === end.getUTCMonth() &&
                   start.getUTCDate() === end.getUTCDate();

    const sameMonth = start.getUTCFullYear() === end.getUTCFullYear() &&
                     start.getUTCMonth() === end.getUTCMonth();

    const sameYear = start.getUTCFullYear() === end.getUTCFullYear();

    if (sameDay) {
        return `${months[start.getUTCMonth()]} ${start.getUTCDate()}, ${start.getUTCFullYear()}`;
    }

    if (sameMonth) {
        return `${months[start.getUTCMonth()]} ${start.getUTCDate()}–${end.getUTCDate()}, ${start.getUTCFullYear()}`;
    }

    if (sameYear) {
        return `${months[start.getUTCMonth()]} ${start.getUTCDate()} – ${months[end.getUTCMonth()]} ${end.getUTCDate()}, ${start.getUTCFullYear()}`;
    }

    return `${months[start.getUTCMonth()]} ${start.getUTCDate()}, ${start.getUTCFullYear()} – ${months[end.getUTCMonth()]} ${end.getUTCDate()}, ${end.getUTCFullYear()}`;
}

// Export for testing
module.exports = { loadWasm, timeago, duration, parse_duration, human_date, date_range };

// Run tests if executed directly
if (require.main === module) {
    const yaml = require('js-yaml');

    async function runTests() {
        await loadWasm();
        console.log('WebAssembly module loaded successfully');

        const testsPath = '/tmp/whenwords/tests.yaml';
        const testsContent = fs.readFileSync(testsPath, 'utf8');
        const tests = yaml.load(testsContent);

        let passed = 0;
        let failed = 0;

        // Run timeago tests
        console.log('\n=== timeago tests ===');
        for (const test of tests.timeago) {
            const result = timeago(test.input.timestamp, test.input.reference);
            if (result === test.output) {
                passed++;
            } else {
                failed++;
                console.log(`FAIL: ${test.name}`);
                console.log(`  timeago(${test.input.timestamp}, ${test.input.reference})`);
                console.log(`  Expected: ${test.output}`);
                console.log(`  Got: ${result}`);
            }
        }

        // Run duration tests
        console.log('\n=== duration tests ===');
        for (const test of tests.duration) {
            const opts = test.input.options || {};
            const result = duration(test.input.seconds, opts.compact || false, opts.max_units || 2);
            if (result === test.output) {
                passed++;
            } else {
                failed++;
                console.log(`FAIL: ${test.name}`);
                console.log(`  duration(${test.input.seconds}, compact=${opts.compact}, max_units=${opts.max_units})`);
                console.log(`  Expected: ${test.output}`);
                console.log(`  Got: ${result}`);
            }
        }

        // Run parse_duration tests
        console.log('\n=== parse_duration tests ===');
        for (const test of tests.parse_duration) {
            const result = parse_duration(test.input);
            if (result === test.output) {
                passed++;
            } else {
                failed++;
                console.log(`FAIL: ${test.name}`);
                console.log(`  parse_duration("${test.input}")`);
                console.log(`  Expected: ${test.output}`);
                console.log(`  Got: ${result}`);
            }
        }

        // Run human_date tests
        console.log('\n=== human_date tests ===');
        for (const test of tests.human_date) {
            const result = human_date(test.input.timestamp, test.input.reference);
            if (result === test.output) {
                passed++;
            } else {
                failed++;
                console.log(`FAIL: ${test.name}`);
                console.log(`  human_date(${test.input.timestamp}, ${test.input.reference})`);
                console.log(`  Expected: ${test.output}`);
                console.log(`  Got: ${result}`);
            }
        }

        // Run date_range tests
        console.log('\n=== date_range tests ===');
        for (const test of tests.date_range) {
            const result = date_range(test.input.start, test.input.end);
            if (result === test.output) {
                passed++;
            } else {
                failed++;
                console.log(`FAIL: ${test.name}`);
                console.log(`  date_range(${test.input.start}, ${test.input.end})`);
                console.log(`  Expected: ${test.output}`);
                console.log(`  Got: ${result}`);
            }
        }

        console.log(`\n=== Results ===`);
        console.log(`Passed: ${passed}`);
        console.log(`Failed: ${failed}`);
        console.log(`Total: ${passed + failed}`);
    }

    runTests().catch(console.error);
}
